<a href="https://www.kaggle.com/code/chiemelachinedum/food-balance-sheets?scriptVersionId=131457441" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/food-data/FoodBalanceSheets_E_Africa_NOFLAG.csv


In [2]:
# import pandas
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# Read the csv file into a variable heart_data
food_data = pd.read_csv("/kaggle/input/food-data/FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding = "latin-1")

In [4]:
display(food_data)

Area Code      Area  Item Code           Item  Element Code  \
0              4   Algeria       2501     Population           511   
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014  \
0               Total Population - Both sexes     1000 persons  38924.00   
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

# Question 1
# Which of these python data structures is unorderly?
Dictionary
List
Tuple
Set

Answer is Set

# Question 2
# What is the total Protein supply quantity in Madagascar in 2015?


In [5]:
food_data_madag = food_data.loc[(food_data.Area == 'Madagascar') & (food_data.Element == 'Protein supply quantity (g/capita/day)')]
food_data_madag['Y2015'].sum()

173.04999999999998

* The total Protein supply quantity in Madagascar in 2015 is 173.05

# Question 3
# Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?


In [6]:
food_data.groupby(['Area'], sort=True)['Y2017'].sum()

Area
Algeria                         325644.27
Angola                          229159.57
Benin                           124771.22
Botswana                         22101.30
Burkina Faso                    101855.07
Cabo Verde                       14650.74
Cameroon                        232030.43
Central African Republic         29937.00
Chad                             71594.68
Comoros                             59.84
Congo                            41181.68
Côte d'Ivoire                   224599.01
Djibouti                         22729.91
Egypt                           866379.92
Eswatini                         54343.33
Ethiopia                        448683.76
Ethiopia PDR                         0.00
Gabon                            27979.64
Gambia                           23154.18
Ghana                           337599.06
Guinea                           98138.87
Guinea-Bissau                    19102.77
Kenya                           264660.66
Lesotho                      

* Capo Verde has the 7th lowest sum in 2017

# Question 4
# What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?


In [7]:
food_data.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

* The mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places are 140.92 and 1671.86

# Question 5
# Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?


In [8]:
food_data.groupby(['Element']).sum()

/tmp/ipykernel_20/1641616304.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  food_data.groupby(['Element']).sum()


Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

* The total number of sum of processing in 2017 is 292836.00

# Question 6
 # Which of the following dataframe methods can be used to access elements across rows and columns?
df.loc( )
df.iloc( )
df.loc( : )
df.iloc[ : ] 
c and d

The answer is df.loc()

# Question 7
# What would be the output for?
S = [['him', 'sell'], [90, 28, 43]]
S[0][1][1]


In [9]:
S = [['him', 'sell'], [90, 28, 43]]
S[0][1][1]


'e'

* The answer is 'e'

# Question 8
# Which year had the least correlation with ‘Element Code’?


In [10]:
food_data.corr()

/tmp/ipykernel_20/3472151187.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  food_data.corr()


Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

* The year 2016 had the least correlation with 'Element Code'

# Question 9
# Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [11]:
food_data.groupby(['Element']).sum()

/tmp/ipykernel_20/1641616304.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  food_data.groupby(['Element']).sum()


Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

* 2014 has the highest sum of Stock Variation


# Question 10
# How would you check for the number of rows and columns in a pandas DataFrame named df?
 len(df)
df.shape( )
pd.length(df)
df.shape

The answer is df.shape()

# Question 11
# Which of the following is a python inbuilt module?
Math
Pandas
Matplotlib
Seaborn

The answer is Math

# Question 12
# Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?


In [12]:
food_data.groupby(['Y2017', 'Area']).sum()

/tmp/ipykernel_20/363971789.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  food_data.groupby(['Y2017', 'Area']).sum()


Area Code  Item Code  Element Code  \
Y2017     Area                                                              
-1582.0   Egypt                              118       5445         10144   
-1200.0   Mozambique                         144       2532          5072   
-1159.0   Mozambique                         144       2907          5072   
-865.0    United Republic of Tanzania        215       2913          5072   
-860.0    United Republic of Tanzania        215       2556          5072   
...                                          ...        ...           ...   
 96443.0  Egypt                               59       2501           511   
 106400.0 Ethiopia                           238       2501           511   
 112625.0 Nigeria                            159       2907          5301   
 117702.0 Nigeria                            159       2907          5511   
 190873.0 Nigeria                            159       2501           511   

                                          Y2014     Y2015     Y2016     Y2018  
Y2017     Area                                                                 
-1582.0   Egypt                          5462.0      94.0   -1970.0   -4312.0  
-1200.0   Mozambique                        0.0       0.0     915.0   -1445.0  
-1159.0   Mozambique                     -219.0      -3.0     915.0   -1442.0  
-865.0    United Republic of Tanzania    2308.0    2505.0    -195.0     -14.0  
-860.0    United Republic of Tanzania     480.0     572.0    -212.0     -61.0  
...                                         ...       ...       ...       ...  
 96443.0  Egypt                         90425.0   92443.0   94447.0   98424.0  
 106400.0 Ethiopia                      98094.0  100835.0  103603.0  109224.0  
 112625.0 Nigeria                      108756.0  110798.0  116483.0  117038.0  
 117702.0 Nigeria                      109593.0  111663.0  119330.0  118172.0  
 190873.0 Nigeria                      176405.0  181137.0  185960.0  195875.0  

[14773 rows x 7 columns]

* Nigeria has the highest sum in 2017

# Question 13
# Given the following python code, what would the output of the code give?
my_tuppy = (1,2,5,8)
my_tuppy[2] = 6
Type Error
Value Error
Standard Error
Math Error


* It produces a Type error

# Question 14
# A pandas Dataframe with dimensions (100,3) has how many features and observations?
100 rows and 3 observations.
A pandas dataframe has no attribute feature or observation.
3 features, 100 observations
100 features, 3 observations.


* It represents 100 rows and 3 observations

# Question 15
# If you have the following list
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]
col = [‘Age’,’Nationality’,’Overall’]
How do you create a pandas DataFrame using this list, to look like the table below?
 


In [13]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]
col = ['Age', 'Nationality', 'Overall']

pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

* This is the answer
pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])

# Question 16
# Answer the following questions based on the African food production dataset provided by the FAO website already provided
# What is the total sum of Wine produced in 2015 and 2018 respectively?
Hint:
Perform a groupby sum aggregation on ‘Item’


In [14]:
food_data.groupby(['Item']).sum()

/tmp/ipykernel_20/3251484073.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  food_data.groupby(['Item']).sum()


Area Code  Item Code  Element Code      Y2014      Y2015  \
Item                                                                            
Alcohol, Non-Food        38447     752497       1515090    2403.00    2180.00   
Alcoholic Beverages      75016    1649136       2171381  102410.11   98783.72   
Animal Products          18060     397035         90990   11935.65   11811.73   
Animal fats              85279    1882494       2567303  209460.54  200675.72   
Apples and products      69280    1352989       1920418    9499.23   10559.15   
...                        ...        ...           ...        ...        ...   
Vegetables, Other        77779    1518715       2275409  155038.96  158104.08   
Vegetal Products         18060     391905         90990  107145.19  107064.17   
Wheat and products       87073    1639683       2654934  232670.13  234710.51   
Wine                     66206    1319535       1818328    4497.36    4251.81   
Yams                     50798     970905       1525745  200396.96  203151.78   

                         Y2016      Y2017      Y2018  
Item                                                  
Alcohol, Non-Food      2223.00    2348.00    2293.00  
Alcoholic Beverages   96958.75   95581.06   97847.27  
Animal Products       11661.69   11547.65   11578.61  
Animal fats          183314.22  269617.53  269648.27  
Apples and products    9853.26   10198.90    9640.51  
...                        ...        ...        ...  
Vegetables, Other    156218.90  157752.59  163987.21  
Vegetal Products     106792.72  107655.20  107775.39  
Wheat and products   239377.07  240047.62  242645.19  
Wine                   3872.09    4178.02    4039.32  
Yams                 220626.20  229174.59  221272.09  

[119 rows x 8 columns]

* Total sum of wine produced in 2015 and 2018 are 4251.81 and 4039.32 respectively

# Question 18
# What is the total number and percentage of missing data in 2014 to 3 decimal places?


In [15]:
food_data.isna().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

* There are 1589 and 2.607% of missing data in 2014

# Question 20
# What is the total number of unique countries in the dataset?

In [16]:
food_data['Area'].unique()

array(['Algeria', 'Angola', 'Benin', 'Botswana', 'Burkina Faso',
       'Cabo Verde', 'Cameroon', 'Central African Republic', 'Chad',
       'Comoros', 'Congo', "Côte d'Ivoire", 'Djibouti', 'Egypt',
       'Eswatini', 'Ethiopia', 'Ethiopia PDR', 'Gabon', 'Gambia', 'Ghana',
       'Guinea', 'Guinea-Bissau', 'Kenya', 'Lesotho', 'Liberia',
       'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mauritius',
       'Morocco', 'Mozambique', 'Namibia', 'Niger', 'Nigeria', 'Rwanda',
       'Sao Tome and Principe', 'Senegal', 'Seychelles', 'Sierra Leone',
       'South Africa', 'Sudan', 'Sudan (former)', 'Togo', 'Tunisia',
       'Uganda', 'United Republic of Tanzania', 'Zambia', 'Zimbabwe'],
      dtype=object)

* There are 50 countries in the dataset
